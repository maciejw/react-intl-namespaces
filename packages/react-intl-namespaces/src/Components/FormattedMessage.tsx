import * as React from 'react';
import * as ReactIntl from 'react-intl';
import { IntlNamespaceContext } from '../context';
import { invariant } from '../invariant';
import { InltNamespaces } from '../namespaces';

@IntlNamespaceContext.Provide
export class FormattedMessage extends ReactIntl.FormattedMessage {
  // prettier-ignore
  public context!: IntlNamespaceContext.Context;
  constructor(
    props: ReactIntl.FormattedMessage.Props,
    context: IntlNamespaceContext.Context,
  ) {
    if (context.intlNamespace === undefined) {
      invariant(
        false,
        'Missing intlNamespace context. Use IntlNamespaceProvider inside IntlBackendProvider',
      );
    }
    super(props, context);
  }

  public render() {
    const result = super.render();

    if (this.context.intlNamespace.includeMetadata) {
      const currentNamespace = this.context.intlNamespace.getNameOfCurrentNamespace();
      const metadata = InltNamespaces.getMessageMetadata(
        this.props,
        currentNamespace,
      );

      return React.createElement(
        'span',
        {
          'data-default-message': metadata.defaultMessage,
          'data-description': metadata.description,
          'data-key': metadata.key,
          'data-ns': metadata.namespace,
        },
        result,
      );
    } else {
      return result;
    }
  }
}
