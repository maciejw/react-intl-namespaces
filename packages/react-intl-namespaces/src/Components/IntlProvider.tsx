import * as React from 'react';
import * as ReactIntl from 'react-intl';
import { IntlNamespaceContext } from '../context';
import { invariant } from '../invariant';
import { IntlNamespaces } from '../namespaces';
import { NamespaceResource } from '../types';

@IntlNamespaceContext.Provide
export class IntlProvider extends ReactIntl.IntlProvider {
  // prettier-ignore
  public context!: IntlNamespaceContext.Context;
  constructor(props: ReactIntl.IntlProvider.Props, context: {}) {
    super(props, context);
  }
  public getChildContext() {
    const result = super.getChildContext();

    const { formatMessage: intlFormatMessage, ...rest } = result.intl;

    const formatMessage = (
      messageDescriptor: ReactIntl.FormattedMessage.MessageDescriptor,
      values?: { [key: string]: ReactIntl.MessageValue },
    ) => {
      const resource: NamespaceResource = this.props.messages || {};
      if (this.context.intlNamespace === undefined) {
        invariant(
          false,
          'Missing intlNamespace context. Use IntlNamespaceProvider inside IntlBackendProvider',
        );
      }

      const {
        getNameOfCurrentNamespace,
        missingMessage,
      } = this.context.intlNamespace;

      if (!resource.hasOwnProperty(messageDescriptor.id)) {
        missingMessage(messageDescriptor);
      }

      if (this.context.intlNamespace.showIds) {
        return IntlNamespaces.getResourceKey(
          messageDescriptor,
          getNameOfCurrentNamespace(),
          Object.getOwnPropertyNames(values),
        );
      } else {
        return intlFormatMessage(messageDescriptor, values);
      }
    };

    return {
      intl: {
        formatMessage,
        ...rest,
      },
    };
  }
  public render() {
    return super.render();
  }
}
