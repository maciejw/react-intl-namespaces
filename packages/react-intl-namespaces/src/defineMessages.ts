import { Messages } from './types';

export function defineMessages<T extends string>(
  messages: Messages<T>,
): Messages<T>;
export function defineMessages(messages: any) {
  return messages;
}
