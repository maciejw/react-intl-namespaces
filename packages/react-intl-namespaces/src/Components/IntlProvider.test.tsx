import * as React from 'react';

import { mount } from 'enzyme';
import { FormattedMessage } from './FormattedMessage';
import { IntlProvider } from './IntlProvider';

import { defineMessages } from '../defineMessages';

const messages = defineMessages<'property1' | 'property2'>({
  property1: {
    defaultMessage: 'text1 {someValue}',
    description: 'description1',
    id: 'property1',
  },
  property2: {
    defaultMessage: 'text2',
    description: 'description2',
    id: 'property2',
  },
});

const values = { someValue: 'value1' };

describe('IntlProvider', () => {
  const sut = (
    <IntlProvider locale="en">
      <div>
        <FormattedMessage {...messages.property1} values={values} />
        <FormattedMessage {...messages.property2} />
      </div>
    </IntlProvider>
  );

  it('should show display message', () => {
    const missingMessage = jest.fn();

    const wrapper = mount(sut, {
      context: {
        intlNamespace: {
          missingMessage,
        },
      },
    });
    expect(wrapper.text()).toMatchSnapshot();
    expect(missingMessage).toBeCalledWith({
      defaultMessage: 'text1 {someValue}',
      description: 'description1',
      id: 'property1',
    });
  });
  it('should show message key when forced', () => {
    const missingMessage = jest.fn();
    const getNameOfCurrentNamespace = jest.fn().mockReturnValue('namespace');

    const wrapper = mount(sut, {
      context: {
        intlNamespace: {
          getNameOfCurrentNamespace,
          missingMessage,
          showIds: true,
        },
      },
    });
    expect(wrapper.text()).toMatchSnapshot();
    expect(missingMessage).toBeCalledWith({
      defaultMessage: 'text1 {someValue}',
      description: 'description1',
      id: 'property1',
    });
    expect(getNameOfCurrentNamespace).toBeCalledWith();
  });
});
