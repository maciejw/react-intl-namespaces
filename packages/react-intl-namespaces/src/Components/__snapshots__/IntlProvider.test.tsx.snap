// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntlProvider should show display message 1`] = `"text1 value1text2"`;

exports[`IntlProvider should show message key when forced 1`] = `"[namespace:property1 (someValue)][namespace:property2 ()]"`;
